### PR TITLE
Retyping: unbound rels raise RetypeError not Not_found

### LINKED ISSUE #13946: Inductive predicate with implicit arguments causes Anomaly
#### Description of the problem

It seems like using implicit arguments in the definition of `elem_of_hlist` causes an `Anomaly Not_found`. I've reduced my test case to the following. I would expect this to function or at the very least return a specific error message. 

```coq
Class ElemOf A B := elem_of: A -> B -> Prop.
Infix "∈" := elem_of (at level 70) .
Inductive Types := tnil: Types | tcons: Type -> Types -> Types.
Implicit Type As: Types.
Infix "^::" := tcons (at level 60, right associativity).
Notation "^[ ]" := tnil (at level 5, format "^[ ]").

Inductive hlist F : Types -> Type :=
| hnil: hlist F ^[]
| hcons {A As} : F A -> hlist F As -> hlist F (A ^:: As).
Infix "+::" := (hcons _) (at level 60, right associativity).

Inductive elem_of_hlist  {F A As} : ElemOf (F A) (hlist F As) :=
  | elem_of_hlist_here F A As (x : F A) (l : hlist F As) : x ∈ (x +:: l)
  | elem_of_list_further F A As (x y : F A) (l : hlist F As) : x ∈ l -> x ∈ (y +:: l).
```

#### Coq Version

8.13

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fix #13946

### CODE REVIEW

**Review by gares — APPROVED** [review]
why not carrying the int as well and print it?

File: pretyping/find_subterm.ml [hunk 1]
```diff
@@ -86,6 +86,7 @@ let replace_term_occ_gen_modulo env sigma occs like_first test bywhat cl occ t =
   let rec substrec k t =
     if Locusops.occurrences_done !count then t else
     try
+      (if not (EConstr.Vars.noccur_between sigma 1 k t) then raise (NotUnifiable None));
```
Comment (SkySkimmer) [inline pretyping/find_subterm.ml:89]: this is probably too violent but let's see if CI can tell

Comment (SkySkimmer) [reply pretyping/find_subterm.ml]: The test suite complexity/pattern example started timing out, let's try putting the check in unification instead.

**Comment by SkySkimmer** [comment]
@coqbot ci minimize

**Comment by SkySkimmer** [comment]
Further reduced
~~~coq

Inductive lookahead_action (term:unit) := Shift_act.

Class IsValidator (P : Prop) (b : bool) := {}.

Global Hint Extern 2 (IsValidator (match ?u with _ => _ end) ?b0) =>
  let b := fresh "b" in
  unshelve notypeclasses refine (let b : bool := _ in _);
  [destruct u; intros; shelve|];
  unify b b0;
  unfold b; destruct u; clear b
  : typeclass_instances.

Axiom Awt : forall t:unit, lookahead_action t.

Lemma foo x0
  : exists g:bool, forall x1,
      IsValidator
        match x0 with
        | tt =>
          match Awt x1 with
          | Shift_act _ => False
          end
        end g.
Proof.
  eexists.
  apply _.
~~~

**Comment by SkySkimmer** [comment]
bench https://gitlab.com/coq/coq/-/jobs/1743677461

**Comment by SkySkimmer** [comment]
~~~
┌─────────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬──────────────────────────┬─────────────────┐
│                             │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]   │   mem faults    │
│                             │                         │                                       │                                       │                          │                 │
│                package_name │     NEW      OLD  PDIFF │            NEW             OLD  PDIFF │            NEW             OLD  PDIFF │     NEW      OLD   PDIFF │ NEW  OLD  PDIFF │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│                    coq-hott │  151.88   153.97  -1.36 │   685438671593    693141436320  -1.11 │  1083281803877   1091201915382  -0.73 │  660928   660632    0.04 │   0    0    nan │
│      coq-mathcomp-ssreflect │   25.46    25.73  -1.05 │   115010983862    116086254052  -0.93 │   146674489593    146665340833   0.01 │  542356   537148    0.97 │   0    0    nan │
│                    coq-core │   96.81    97.82  -1.03 │   392039547005    388470277303   0.92 │   435367612785    435332328273   0.01 │  248096   248884   -0.32 │   0    0    nan │
│              coq-coquelicot │   34.21    34.54  -0.96 │   154012936137    154526725461  -0.33 │   203621287317    203599975457   0.01 │  747360   749400   -0.27 │   0    0    nan │
│           coq-iris-examples │  453.33   456.59  -0.71 │  2056380366628   2070951973224  -0.70 │  3127873028543   3131939691778  -0.13 │ 1178952  1182184   -0.27 │   0    0    nan │
│       coq-mathcomp-solvable │   88.89    89.38  -0.55 │   405184564616    407671440078  -0.61 │   630178898368    630235638889  -0.01 │  663468   664748   -0.19 │   0    0    nan │
│                coq-rewriter │  373.93   375.70  -0.47 │  1693931799201   1700603197227  -0.39 │  2787638804040   2791827110840  -0.15 │ 1063236  1067140   -0.37 │   0    0    nan │
│               coq-fourcolor │ 1506.07  1511.92  -0.39 │  6660984047598   6682451792040  -0.32 │ 12104829171733  12104530831245   0.00 │  730460   730604   -0.02 │   0    0    nan │
│               coq-fiat-core │   53.51    53.69  -0.34 │   231643740039    231303592928   0.15 │   323141337913    323143895808  -0.00 │  474120   476228   -0.44 │   0    0    nan │
│                  coq-geocoq │  712.41   714.33  -0.27 │  3240462539453   3246142005328  -0.17 │  5158178198835   5152418751349   0.11 │ 1057232  1047152    0.96 │   0    0    nan │
│                   coq-verdi │   48.42    48.54  -0.25 │   219456288323    218656763943   0.37 │   326828983662    325926389852   0.28 │  830692   831612   -0.11 │   0    0    nan │
│      coq-mathcomp-odd-order │  568.77   570.11  -0.24 │  2601611695249   2607117120349  -0.21 │  4475552311402   4475825729180  -0.01 │  938544   938208    0.04 │   0    0    nan │
│ coq-rewriter-perf-SuperFast │  897.50   899.04  -0.17 │  4040618675052   4045305414002  -0.12 │  6968841712158   6972707984688  -0.06 │ 1147264  1148128   -0.08 │   0    0    nan │
│                 coq-bignums │   28.21    28.25  -0.14 │   128019073603    128407402483  -0.30 │   176489007880    176621983862  -0.08 │  467796   468000   -0.04 │   0    0    nan │
│  coq-performance-tests-lite │  781.01   782.04  -0.13 │  3510529671545   3513342345534  -0.08 │  5985336430102   6016379791186  -0.52 │ 1833104  2062916  -11.14 │   0    0    nan │
│                   coq-color │  258.78   259.06  -0.11 │  1165676053427   1165823587894  -0.01 │  1590665855653   1591894871207  -0.08 │ 1241408  1241436   -0.00 │   0    0    nan │
│        coq-mathcomp-algebra │   63.50    63.52  -0.03 │   288920849309    289222935437  -0.10 │   404221080600    404173632912   0.01 │  562024   561688    0.06 │   0    0    nan │
│                 coq-unimath │ 2688.39  2685.29   0.12 │ 12229178149877  12218097524252   0.09 │ 24348935814386  24344385541965   0.02 │ 3837376  3837236    0.00 │   0    0    nan │
│          coq-mathcomp-field │  115.40   115.24   0.14 │   526770499802    526427771664   0.07 │   873336111127    873418433372  -0.01 │  660792   660852   -0.01 │   0    0    nan │
│               coq-perennial │ 3825.52  3820.02   0.14 │ 17358577686096  17325766109205   0.19 │ 28888883876894  28846629831078   0.15 │ 2691884  2712536   -0.76 │   0    0    nan │
│                    coq-corn │  792.33   791.17   0.15 │  3598651918271   3591415948265   0.20 │  5605804816366   5605195898888   0.01 │  829268   828340    0.11 │   0    0    nan │
│       coq-mathcomp-fingroup │   23.44    23.38   0.26 │   106769049612    106486478580   0.27 │   156942788847    156963656228  -0.01 │  481624   481060    0.12 │   0    0    nan │
│                   coq-flocq │   74.44    74.24   0.27 │   337192592196    337471021483  -0.08 │   448048212595    448011625153   0.01 │  994300   992132    0.22 │   0    0    nan │
│      coq-mathcomp-character │   77.43    77.15   0.36 │   353156211131    352440563135   0.20 │   540292566829    540271241217   0.00 │  742852   738004    0.66 │   0    0    nan │
│                 coq-coqutil │   31.86    31.74   0.38 │   142449591023    141270955933   0.83 │   188127494534    188117741205   0.01 │  600032   600820   -0.13 │   0    0    nan │
│             coq-fiat-crypto │ 2787.06  2776.41   0.38 │ 12572334138369  12525315056566   0.38 │ 23224565601431  23151593353248   0.32 │ 2251560  2266468   -0.66 │   0    0    nan │
│            coq-fiat-parsers │  332.29   330.66   0.49 │  1476257433564   1468792303045   0.51 │  2433059359182   2431879523956   0.05 │ 2799612  2798692    0.03 │   0    0    nan │
│              coq-verdi-raft │  577.91   574.79   0.54 │  2631041629384   2618246674963   0.49 │  4049384871067   4037049767490   0.31 │ 1214788  1219340   -0.37 │   0    0    nan │
│            coq-math-classes │   92.33    91.81   0.57 │   416782043953    416214661099   0.14 │   590157564783    590067275760   0.02 │  495660   495036    0.13 │   0    0    nan │
│       coq-engine-bench-lite │  176.75   175.73   0.58 │   757799817801    753583784282   0.56 │  1477829624149   1476526403122   0.09 │ 1294304  1294364   -0.00 │   0    0    nan │
│                coq-coqprime │  178.66   177.39   0.72 │   815068247807    808838448745   0.77 │  1562046045112   1561377771706   0.04 │  831932   858940   -3.14 │   0    0    nan │
│                  coq-stdlib │  412.04   408.68   0.82 │  1687910483904   1669869308179   1.08 │  1445995665970   1446493441141  -0.03 │  598076   594480    0.60 │   0    0    nan │
│                coq-bedrock2 │  172.64   170.70   1.14 │   775510703815    766123819656   1.23 │  1450144788919   1431261872269   1.32 │  789980   791212   -0.16 │   0    0    nan │
└─────────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴──────────────────────────┴─────────────────┘
~~~

File: pretyping/find_subterm.ml [hunk 2]
```diff
@@ -86,6 +86,7 @@ let replace_term_occ_gen_modulo env sigma occs like_first test bywhat cl occ t =
   let rec substrec k t =
     if Locusops.occurrences_done !count then t else
     try
+      (* (if not (EConstr.Vars.noccur_between sigma 1 k t) then raise (NotUnifiable None)); *)
```
Comment (gares) [inline pretyping/find_subterm.ml:89]: Can you remove this before merging?

**Comment by gares** [comment]
@coqbot merge now

**Pull request merged**